src/mon/OSDMonitor.cc

```diff
@@ -7687,37 +7687,41 @@ int OSDMonitor::get_crush_rule(const string &rule_name,
   }
   return 0;
 }
-
-int OSDMonitor::check_pg_num(int64_t pool, int pg_num, int size, ostream *ss)
+int OSDMonitor::check_pg_num(int64_t pool, int pg_num, int size, int rule, ostream *ss)
 {
   auto max_pgs_per_osd = g_conf().get_val<uint64_t>("mon_max_pg_per_osd");
-  auto num_osds = std::max(osdmap.get_num_in_osds(), 3u);   // assume min cluster size 3
-  auto max_pgs = max_pgs_per_osd * num_osds;
   uint64_t projected = 0;
-  if (pool < 0) {
-    projected += pg_num * size;
-  }
-  for (const auto& i : osdmap.get_pools()) {
-    if (i.first == pool) {
-      projected += pg_num * size;
-    } else {
-      projected += i.second.get_pg_num_target() * i.second.get_size();
-    }
-  }
+  unsigned osd_num = 0;
+  CrushWrapper newcrush;
+  _get_pending_crush(newcrush);
+  set<int> roots;
+  newcrush.find_takes_by_rule(rule, &roots);
+  for (auto root : roots) {
```

Comment:
i don't think this is right. actually, this might be much less accurate than the existing evaluation algorithm.

what if the cluster has existing pools? and what if osds picked by rules used by those pools are overlapped the ones used by this rule? we need to take them into consideration.

Reply:
As I understand it, get_osd_acting_pgs can get the number of pgs on the current osd, in fact these pgs contain pgs from pools already in the cluster, so the total number of pgs calculated using this is the total number of pgs picked by rules used by ossds on this pool. This pg count should be accurate.

Reply:
@fyzard1991 thanks. now i get it. we only care about the number of osds assigned to `pool`. so it does not matter if a certain osd is also chosen by an existing pool or not.